pkg/planner/property/physical_property.go

```diff
@@ -297,6 +305,75 @@ func (p *PhysicalProperty) IsSubsetOf(keys []*MPPPartitionColumn) []int {
 	return matches
 }
 
+// NeedEnforceExchangerWithHashByEquivalence checks if the keys can match the needs of partition with equivalence.
+// "Equivalence" refers to the process where we utilize a hash column to obtain equivalent columns,
+// and then use these equivalent columns to compare with the MPP partition column to determine whether an exchange is
+// necessary.
+//
+// for example:
+//  1. MPPPartitionCols: [18，13，16]
+//  2. keys: [9]
+//  3. FD: (1)-->(2-6,8), ()-->(7), (9)-->(10-17), (1,10)==(1,10), (18,21)-->(19,20,22-33), (9,18)==(9,18)
+//     In this case, we can see that the child supplied partition keys is subset of parent required partition cols.
+func (p *PhysicalProperty) NeedEnforceExchangerWithHashByEquivalence(keys []*MPPPartitionColumn) bool {
+	// keys is the HashCol. If the partition cols are a subset of the hash cols, then need to enforce exchange.
+	if len(p.MPPPartitionCols) < len(keys) {
+		return true
+	}
+	uniqueID2MppCol := make(map[*MPPPartitionColumn]intset.FastIntSet, len(keys))
+	// for each partition column, we calculate the equivalence alternative closure of it.
+	for _, pCol := range p.MPPPartitionCols {
+		uniqueID2MppCol[pCol] = p.FD.ClosureOfEquivalence(intset.NewFastIntSet(int(pCol.Col.UniqueID)))
+	}
+
+	// there is a subset theorem here, if the child supplied keys is a subset of parent required mpp partition cols,
+	// the mpp partition exchanger can also be eliminated.
+	isSubset := true
+SubsetLoop:
+	for _, key := range keys {
+		for pCol, equivSet := range uniqueID2MppCol {
+			if checkEquivalence(equivSet, key, pCol) {
+				// yes, child can supply the same col partition prop. continue to next child supplied key.
+				continue SubsetLoop
+			}
+		}
+		// once a child supplied keys can't find direct/in-direct equiv all parent required partition cols,
+		// we can break the subset check.
+		isSubset = false
+		break SubsetLoop
+	}
+	// it's subset case, we don't need to add exchanger.
+	if isSubset {
+		return false
+	}
+
+	// when arrived here, it means the child supplied partition cols is not subset of parent required partition cols.
+	// then we need to check if the child supplied partition cols can supply the every partition col as parent required.
+	// iterate the parent prop required partition cols, check if it is compatible with children supplied.
+NextRequiredCol:
+	for pCol, equivSet := range uniqueID2MppCol {
+		// iterate the child supplied keys.
+		for _, key := range keys {
+			if checkEquivalence(equivSet, key, pCol) {
+				// yes, child can supply the same col partition prop. continue out to next required partition col.
+				continue NextRequiredCol
```

Comment:
This continue will only skip the inner side loop which means that it will always goto the line:362 and return true 。。。。。。